### PR TITLE
✨ Add implementation of Node IPAM support

### LINKED ISSUE #1550: IPAM support for CAPV
Umbrella issue for all tasks required to implement IPAM support for CAPV
This issue will be updated over time

- [ ] IPAM Contract implementation - https://github.com/kubernetes-sigs/cluster-api/pull/6313

### PR DESCRIPTION
**Pull request by tylerschultz**
**What this PR does / why we need it**:

- Adds `addressesFromPools` to the VSphereVM network device spec
- Creates IPAddressClaims when AddressesFromPools is specified
- IP addresses created by IPAM providers are assigned to vSphere machines via the VM metadata
- IPv4 Gateways from IPAddresses (created by IPAM providers) are assigned to the Gateway4 field
- IPv6 Gateways from IPAddresses (created by IPAM providers) are assigned to the Gateway6 field
- All gateways in the same IP family assigned to a VSphereVM network device must be the same.
- This implementation expects that an appropriate gateway for the IP family exists for every IPAddress that comes from an IPAM provider.
- Adds the `IPAddressClaimed` Condition to provide visibility of the status of IP address acquisition
- Adds the finalizer to prevent IPAMClaims from being deleted early
- IPAddressClaims have an OwnerReference to VSphereVMs to ensure they are garbage-collected when the VM is deleted.

Co-authored-by: Aidan Obley <aobley@vmware.com>
Co-authored-by: Tyler Schultz <tschultz@vmware.com>
Co-authored-by: Christian Ang <angc@vmware.com>
Co-authored-by: Edwin Xie <exie@vmware.com>

**Which issue(s) this PR fixes**:
Fixes #1550 

This PR implements this proposal: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1632

**Special notes for your reviewer**:

The e2e test will come in a separate PR. The IPAM feature requires an unreleased version of CAPI with a version of clusterctl with `--ipam-provider` flag and related IPAM CRDs. The main branch of CAPI also bumps to ginkgo v2. https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1667

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Adds support for node IPAM
```

### CODE REVIEW

File: controllers/vspherevm_controller.go
Comment (tylerschultz) [inline controllers/vspherevm_controller.go:326]: This line should be called out. We see that other errors in the delete reconcile do not return errors, they log and eat the errors. We've chosen to return the error during finalizer removal errors.

Does anyone have context on why errors are not returned? Should this new code not return errors?

File: pkg/util/machines.go
Comment (srm09) [inline pkg/util/machines.go:122]: Update the comment and consider making it a private method if not used directly outside the package.

File: pkg/util/machines_test.go
Comment (srm09) [inline pkg/util/machines_test.go:651]: nit: describe the case better

File: apis/v1beta1/vspherevm_types.go
Comment (srm09) [inline apis/v1beta1/vspherevm_types.go:34]: What do you think about changing this reflect that it is being controller by CAPV, 
`vspherevm.infrastructure.cluster.x-k8s.io/ip-claim-protection` ?

**Review by srm09 — COMMENTED** [review]
Did an initial pass, bunch of nits apart from setting the Conditions on the VSphereVM object. Need to fully review the `reconcileIPAddresses` function once, but the changes overall look good.

File: config/rbac/role.yaml
Comment (tylerschultz) [reply config/rbac/role.yaml:321]: Delete does not appear to be necessary. We've seen the objects get deleted successfully. We'll make another pass to be sure.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:337]: While possible, I don't think overwriting the condition is a likely scenario.

The VM is what is being reconciled, so the controller would see all of the desired claims in one pass. If any claims are missing, they'd be created and the status would be set as above.

The only way the status would be overwritten would be if the VM had failed reconciliation in a previous pass, and then  `addressesFromPools` section were to be amended. IIRC, the `VSphereVM`s are not mutated, and instead are recreated.

Per our offline conversation, we said we'd improve the condition status logic to include more information about desired count/actual count in a later PR.

File: controllers/vspherevm_controller.go
Comment (tylerschultz) [reply controllers/vspherevm_controller.go:326]: @srm09 Do you have context on why during delete that errors are ignored and not returned?

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:133]: Correct. Ensure the claims exist. Don't wait for the IPAMProvider to create IPAddresses. The next function waits.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:275]: Sure, I'll add some logging. 

I worry that might be noisy? Each reconcile loop would repeat the found status. The next block of code logs the creation of each claim.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:353]: We had been logging errors, but removed them because we were seeing duplicate logs. 
We're seeing that errors returned during reconciliation are logged. Are there cases where the errors would not be logged?

File: pkg/util/machines.go
Comment (tylerschultz) [reply pkg/util/machines.go:122]: We get apidiff failures when we reduce the visibility of functions in this package. Is that ok? I don't think the functions in this package are used widely. I'm not a big fan of the factoring. The functions are public so that other packages can use them, but they're really internal functions.

File: apis/v1beta1/vspherevm_types.go
Comment (tylerschultz) [reply apis/v1beta1/vspherevm_types.go]: updated.

File: pkg/util/machines_test.go
Comment (tylerschultz) [reply pkg/util/machines_test.go]: updated

File: pkg/manager/manager.go
Comment (tylerschultz) [reply pkg/manager/manager.go]: updated

File: controllers/vspherevm_controller.go
Comment (randomvariable) [reply controllers/vspherevm_controller.go:326]: It depends on whether or not the resource should be immediately requeued for retry. If so, return the error. If you only want to retry during a normal reconciliation interval, eat the error and return nil.

Comment (tylerschultz) [reply controllers/vspherevm_controller.go:326]: In our case, we need to make sure the claim's finalizer is removed, or the IP will not be returned to the pool, so we think it's appropriate to return an error.

It is odd to us that when deleting the VM object that if something were to go wrong, it would eat the error. That is why we ask if there's some other reason we should eat the error too.

Comment (randomvariable) [reply controllers/vspherevm_controller.go:326]: I don't think so. Probably historical mostly and trying to avoid early exponential backoff.

File: pkg/services/govmomi/service.go
Comment (srm09) [reply pkg/services/govmomi/service.go:275]: Higher verbosity, ideally > 4 since CAPV by default uses 4 in its manifests.

File: pkg/util/machines.go
Comment (srm09) [reply pkg/util/machines.go:122]: I meant update the first word in the comment to correctly reflect the name of the method
~~GetMachineMetadata~~ --> GetMachineMetadataWithIPAMState

Comment (tylerschultz) [reply pkg/util/machines.go]: Ahhh, didn't see it. Fixed!

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:275]: updated

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: updated

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: updated

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: updated

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: updated

**Comment by tylerschultz** [comment]
Apologies, the conversations were folded and I didn't see several items! All should be resolved now. Thanks!

File: controllers/vspherevm_controller.go
Comment (srm09) [inline controllers/vspherevm_controller.go:311]: If this controller is responsible for querying and removing the finalizer on the `IPAdressClaim`, then the kubebuilder brace directives should be a part of this file as well. Any particular reason why these get set on the `vspheremachine_controller.go` only.
My vote would be to duplicate for code readability if it is needed in both places. That way each controller clearly states all the rbac permissions it requires.

File: pkg/services/govmomi/service.go
Comment (srm09) [inline pkg/services/govmomi/service.go:309]: nit: remove empty line

File: pkg/services/govmomi/service.go
Comment (srm09) [inline pkg/services/govmomi/service.go:313]: nit: add a comment stating the intent of the method

**Review by srm09 — COMMENTED** [review]
lgtm apart from a few nits.
/approve
/hold

**Comment by srm09** [comment]
@yastij @schrej Do take a pass at reviewing the PR and feel free to remove the hold once you feel it is good to go.

File: controllers/vspherevm_controller.go
Comment (tylerschultz) [reply controllers/vspherevm_controller.go:313]: Good catch, this was a mistake. When we started this work we assumed we'd be modifying the `vspheremachine_controller.go`, but it turned out we needed to modify the `vspherevm_controller.go` instead. We've moved the kubebuilder brace directives to `vspherevm_controller.go`.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: updated

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:319]: updated

File: pkg/services/govmomi/service.go
Comment (schrej) [inline pkg/services/govmomi/service.go:787]: FYI: There is a new `netip` package since go1.18 with a nicer API: https://pkg.go.dev/net/netip#Addr.Is4

**Review by schrej — APPROVED** [review]
Very clean, well done!
Just two small things, otherwise
/lgtm

I didn't look at tests in detail though.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go]: cool, we've switched over to `netip`

File: controllers/vspherevm_controller.go
Comment (tylerschultz) [reply controllers/vspherevm_controller.go]: We missed this one! Thanks

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:137]: We considered putting it sooner, but it seems worthwhile to ensure the devices are in order before creating claims. The checking of network device status seems to run fairly quick - at least in our envs. 

We don't understand what the implications of checking hardware version are, we again thought we should wait for this 'pre-check'.

We like where it is, but we open to moving it to be sooner. Do you have a suggestion? We will be back again soon in this area of the code when we make updates to how IPAM condition status is reported. Perhaps we can revisit then?

File: pkg/util/machines.go
Comment (yastij) [inline pkg/util/machines.go:117]: I don't think we need rename this. ultimately this function is supposed to return the metadata of a VM, we shouldn't leak the content into the name of the function

File: pkg/services/govmomi/service.go
Comment (yastij) [reply pkg/services/govmomi/service.go:137]: +1 I don't think it's worth moving up for now, the sequencing we likely want:
- clone the VM with the devices
- verify devices and report the first wave of info in the status
- reconcile IPClaims
- verify the claims are fulfilled
- reconfigure the VM by setting the metadata

@tylerschultz - the only question I have is mostly around speed, did you notice longer cluster creation time or is it roughly stable?

**Review by yastij — COMMENTED** [review]
/approve

**Comment by yastij** [comment]
lgtm pending answers to the question above

File: pkg/services/govmomi/service.go
Comment (schrej) [reply pkg/services/govmomi/service.go:137]: From a logical perspective that sequence makes sense of course. My argument would be to start with the IP Claim creation. Then clone the VM and verify devices. Then verify that claims are fulfilled.

This just gives the ipam providers a bit of time to potentially fulfil the claims in the meantime. If you create them and then check them right away it's quite likely that they aren't fulfilled yet, and reconciliation is aborted and needs to start over.

But I'm fine with both strategies, was just a suggestion.

File: pkg/util/machines.go
Comment (yastij) [inline pkg/util/machines.go:131]: this might not work with clusterClass, this is a case of co-authored slices. we need to be able a unique key so that server-side apply works. cc @fabriziopandini

**Review by yastij — COMMENTED** [review]
/hold

File: pkg/util/machines.go
Comment (fabriziopandini) [reply pkg/util/machines.go:131]: Confirmed.
if order to allow one contreller (api) to create the device entry, and another (capv) to set the ipAddr into device entries, the list of devices must be defined as +ListMap,

Comment (yastij) [reply pkg/util/machines.go:131]: ok, just confirmed it seems like we're never retrofitting `.spec.devices.ipAddrs` so it shouldn't be an issue. 

@tylerschultz  can you confirm we're not patching `.spec.devices.ipAddrs` of the vsphereVM with the result of the fulfilled claim?

File: pkg/util/machines.go
Comment (tylerschultz) [reply pkg/util/machines.go:117]: We made a second function to preserve the original function. `make apidiff` fails when we add a parameter to the original function. We can add the parameter to the original function name, but that means changing the 'public' api. 
Is it acceptable to submit this PR with failing apidiff?

File: pkg/util/machines.go
Comment (tylerschultz) [reply pkg/util/machines.go:131]: We had originally updated `spec.devices.ipAddrs` on the VSphereVM, but we realized we were modifying the spec of the object so we backed it out. We changed it to use the in-memory map instead. We do not modify the `.spec.devices.ipAddrs`.

File: pkg/services/govmomi/service.go
Comment (tylerschultz) [reply pkg/services/govmomi/service.go:137]: We've not noticed any delay. 

Our take is that in the grand scheme of standing up a cluster, the extra reconcile loop or 2 is trivial compared to the time it takes to boot a vm. OTOH, every loop counts, and perhaps we can optimize this.

We propose leaving where it is now and seeing how it goes.

File: pkg/util/machines.go
Comment (yastij) [reply pkg/util/machines.go:117]: yeah we're using it mostly as a signal to know what we're breaking, we're not providing contract stability for functions

File: pkg/services/govmomi/service.go
Comment (yastij) [reply pkg/services/govmomi/service.go:137]: @schrej - thinking more about, I think creating the IPClaims as part of the controller process before cloning makes sense. Ideally the VM Service layer should be mostly reading resources and translating them into vSphere constructs.

@tylerschultz do you mind filing an issue to address this later?

Comment (tylerschultz) [reply pkg/services/govmomi/service.go:137]: will do

**Comment by srm09** [comment]
/unhold
I will let @yastij add the final lgtm which should let this one merge.

**Comment by tylerschultz** [comment]
Thanks all!

**Pull request merged**
